sdk/protocol2.go

```diff
@@ -0,0 +1,164 @@
+type Entity struct {
+	Name string `json:"name"`
+	Type string `json:"type"`
+}
+
+// EntityData defines all the data related to a particular event from an entity.
+type EntityData struct {
+	Entity    Entity             `json:"entity"`
+	Metrics   []metric.MetricSet `json:"metrics"`
+	Inventory Inventory          `json:"inventory"`
+	Events    []Event            `json:"events"`
+}
+
+// IntegrationProtocol2 defines the format of the output JSON that integrations will return for protocol 2.
+type IntegrationProtocol2 struct {
+	sync.Mutex
+	Name               string        `json:"name"`
+	ProtocolVersion    string        `json:"protocol_version"`
+	IntegrationVersion string        `json:"integration_version"`
+	Data               []*EntityData `json:"data"`
+	prettyOutput       bool
+	writer             io.Writer
+}
+
+// NewIntegrationProtocol2WithWriter initializes a new instance of the integration protocol 2 specifying a writer instead of using stdout by default.
+func NewIntegrationProtocol2WithWriter(name string, version string, arguments interface{}, writer io.Writer) (*IntegrationProtocol2, error) {
+	i, err := NewIntegrationProtocol2(name, version, arguments)
+	if err != nil {
+		return nil, err
+	}
+
+	i.writer = writer
+
+	return i, nil
+}
+
+// NewIntegrationProtocol2 initializes a new instance of the integration protocol 2.
+func NewIntegrationProtocol2(name string, version string, arguments interface{}) (*IntegrationProtocol2, error) {
+	err := args.SetupArgs(arguments)
+	if err != nil {
+		return nil, err
+	}
+	defaultArgs := args.GetDefaultArgs(arguments)
+
+	log.SetupLogging(defaultArgs.Verbose)
+
+	// Avoid working with an uninitialized or in error state cache
+	if err = cache.Status(); err != nil {
+		return nil, err
+	}
+
+	integration := &IntegrationProtocol2{
+		Name:               name,
+		ProtocolVersion:    "2",
+		IntegrationVersion: version,
+		prettyOutput:       defaultArgs.Pretty,
+		writer:             os.Stdout, // StdOut by default
+	}
+
+	return integration, nil
+}
+
+// Entity method creates or retrieves an already created EntityData.
+func (integration *IntegrationProtocol2) Entity(entityName, entityType string) (*EntityData, error) {
+	if entityName == "" || entityType == "" {
+		return nil, errors.New("entityName and entityType should be not empty")
+	}
+
+	integration.Lock()
+	defer integration.Unlock()
+	for _, e := range integration.Data {
+		if e.Entity.Name == entityName && e.Entity.Type == entityType {
+			return e, nil
+		}
+	}
+
+	d := &EntityData{
+		Entity: Entity{
+			Name: entityName,
+			Type: entityType,
+		},
+	}
+
+	integration.Data = append(integration.Data, d)
+
+	return d, nil
+}
+
+// NewMetricSet returns a new instance of MetricSet with its sample attached to
+// the IntegrationData.
+func (d *EntityData) NewMetricSet(eventType string) metric.MetricSet {
+	ms := metric.NewMetricSet(eventType)
```

Comment:
Setting this eventType here is problematic. Maybe the user want to initilize the metricsSet without the eventType and add it later. 
Maybe we could have one more method for `*EntityData` without setting the `eventType`